R/MsExperiment.R

```diff
@@ -214,48 +214,99 @@ setMethod("readMsObject", signature(object = "MsExperiment",
-              merged_data <- merged_data[order(merged_data$injection_index), ]
-              merged_data <- merged_data[, colSums(is.na(merged_data)) <
-                                             nrow(merged_data)]
+              s_files <- all_fls[grepl("^s_", all_fls)]
+              sample_info <- read.table(paste0(pth, s_files),
+                                        header = TRUE, sep = "\t",
+                                        check.names = FALSE)
+
+              # merging
+              ord <- match(assay_data$`Sample Name`, sample_info$`Sample Name`)
+              merged_data <- cbind(assay_data, sample_info[ord, ])
+              names(merged_data) <- gsub(" ", "_", names(merged_data))
+              if (keepProtocol || keepOntology || simplify)
+                  merged_data <- .clean_merged(x = merged_data,
+                                               keepProtocol = keepProtocol,
+                                               keepOntology = keepOntology,
+                                               simplify = simplify)
+
+              ## Assemble object
+              object@spectra <- Spectra::Spectra(mtblsId = param@mtblsId,
+                                                 source = MsBackendMetaboLights::MsBackendMetaboLights(),
+                                                 assayName = selected_assay,
+                                                 filePattern = param@filePattern)
+
+              ## sample to spectra link
+              fl <- object@spectra@backend@spectraData[1, "derived_spectral_data_file"]
+              nme <- colnames(merged_data)[which(merged_data[1, ] == fl)]
+              merged_data <- merged_data[grepl(param@filePattern,
+                                               merged_data[, nme]), ]
 
               object@sampleData <- DataFrame(merged_data)
+              object <- MsExperiment::linkSampleData(object,
+                                                     with = paste0("sampleData.",
+                                                                nme,
+                                                                "= spectra.derived_spectral_data_file"))
               validObject(object)
               object
           })
+
+
+#####HELPERS
+
+#' function that takes the extra parameters and clean the metadata if asked by
+#' the user.
+#'
+#' Note:  the subsetting of the merged data is done here, which WILL rename the
+#' duplicated columns. I could fix that by first transforming the data.frame into
+#' a list but I am not sure that it is useful.. The user might do some
+#' subsetting too later and then the same thing will happen. Might as well have
+#' it from the beginning.
+#'
+#' Note2: I would move that function later, keeping it her for review to help
+#' clarity
+#'
+#' @noRd
+.clean_merged <- function(x, keepProtocol, keepOntology, simplify) {
+    # remove ontology
+    if (!keepOntology)
+        x <- x[, -which(grepl("Term", names(x)))]
```

Comment:
@jorainer continuing the discussion about checking column names or not here. I believe we cannot give the choice to the user whether or not to have numbered column. because as soon as this (or the one below) subsetting happen, there is an automatic renaming of duplicated columns. I could not find a way to prevent that.. 

Also I don't really want to allow the read.table(check.names = TRUE) when importing the assay and sample info, because in each there is already duplication of column, so after merging AND subsetting there would be double numbering... 

Also the tables will be different compared to MetaboLights anyway because a lot of columns are not even present on the MetaboLights website...(they derive from the ontology term that the user input)

Lastly, even If i find a way to have  as an output a table with duplicated column names with NO numbering, if the user subset the `MsExperiment` object by calling `[]` the numbering will get automatically added..

So I think it is a bit more complicated than expected this numbering thing....